blocks/bricks/interfaces.py

```diff
@@ -127,45 +129,121 @@ class Initializable(RNGMixin, Brick):
+                                     "is not defined by default, pass it"
+                                     "via initialization_schemes")
+                if initialization_to_role[key] in \
+                        self.initialization_schemes.keys():
+                    raise ValueError("All initializations are accepted either"
+                                     "through initialization schemes or "
+                                     "corresponding attribute but not both")
+                else:
+                    self.initialization_schemes[initialization_to_role[
+                                                key]] = kwargs[key]
+                kwargs.pop(key)
+
+        super(Initializable, self).__init__(**kwargs)
+
+    def get_scheme(role, schemes):
+        for key in schemes:
+            if role == type(key):
+                return key
+        for key in schemes:
+            if isinstance(role, type(key)):
+                return key
+
+    def _validate_roles(self):
+        all_parent_roles = []
+        for role in self.parameter_roles:
+            all_parent_roles += list(inspect.getmro(type(role)))
+
+        for key in self.initialization_schemes:
+            if type(key) not in all_parent_roles:
+                raise ValueError("There is no parameter role"
+                                 "for initlization sheme {}".format(key))
 
     def _push_initialization_config(self):
+        self._collect_roles()
+        self._validate_roles()
         for child in self.children:
-            if isinstance(child, Initializable):
+            if (isinstance(child, Initializable) and
+                    hasattr(child, 'initialization_schemes')):
                 child.rng = self.rng
-                if self.weights_init:
-                    child.weights_init = self.weights_init
-        if hasattr(self, 'biases_init') and self.biases_init:
-            for child in self.children:
-                if (isinstance(child, Initializable) and
-                        hasattr(child, 'biases_init')):
-                    child.biases_init = self.biases_init
+                for role, scheme in self.initialization_schemes.items():
+                    if role in child.parameter_roles:
+                        child.initialization_schemes[role] = scheme
+
+    def _collect_roles(self):
+        def get_param_roles(obj):
+            all_roles = []
+            for param in obj.parameters:
+                roles = param.tag.roles
+                # TODO do something smarter
+                if len(roles) > 0:
+                    all_roles.append(roles[0])
+            return all_roles
+
+        self.parameter_roles = set(get_param_roles(self))
+        for child in self.children:
+            if isinstance(child, Initializable):
+                child._collect_roles()
+                self.parameter_roles.update(child.parameter_roles)
+
+    def _initialize(self):
+        def get_scheme(role, schemes):
```

Comment:
Right, but all I am using is "isinstance", so what's wrong?

On 6 June 2016 at 16:44, Mehdi Mirza <notifications@github.com> wrote:

> In blocks/bricks/interfaces.py
> <https://github.com/mila-udem/blocks/pull/1030#discussion_r65965285>:
>
> > +            all_roles = []
> > +            for param in obj.parameters:
> > +                roles = param.tag.roles
> > +                # TODO do something smarter
> > +                if len(roles) > 0:
> > +                    all_roles.append(roles[0])
> > +            return all_roles
> > +
> > +        self.parameter_roles = set(get_param_roles(self))
> > +        for child in self.children:
> > +            if isinstance(child, Initializable):
> > +                child._collect_roles()
> > +                self.parameter_roles.update(child.parameter_roles)
> > +
> > +    def _initialize(self):
> > +        def get_scheme(role, schemes):
>
> I'm not sure your code snippet do that either. Let's assume we have this
> class hierarchy A->B->C
> a = A(), b=B(), c=C()
> Then:
> isinstance(c, A) -- True
> type(c) == A -- False
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mila-udem/blocks/pull/1030/files/f3fca5343847c265abe7b7c2424d585485476b76#r65965285>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAn8Yul0FN9mBXfoNDAvqFDqOW5JOviZks5qJIbJgaJpZM4HxS-w>
> .
>